44.md

```diff
@@ -0,0 +1,167 @@
+
+const utf8Encoder = new TextEncoder()
+
+export const getSharedSecret = (privkey: string, pubkey: string): Uint8Array =>
+  sha256(secp256k1.getSharedSecret(privkey, '02' + pubkey).subarray(1, 33))
+
+export function encrypt(key: Uint8Array, text: string, v = 1) {
+  if (v !== 1) {
+    throw new Error('NIP44: unknown encryption version')
+  }
+
+  const nonce = randomBytes(24)
+  const plaintext = utf8Encoder.encode(text)
+  const ciphertext = xchacha20(key, nonce, plaintext)
+
+  const payload = new Uint8Array(25 + ciphertext.length)
+  payload.set([v], 0)
+  payload.set(nonce, 1)
+  payload.set(ciphertext, 25)
+
+  return base64.encode(payload)
+}
+
+export function decrypt(key: string, payload: string) {
+  let data
+  try {
+    data = base64.decode(payload)
+  } catch (e) {
+    throw new Error(`NIP44: failed to base64 decode payload: ${e}`)
+  }
+
+  if (data[0] !== 1) {
+    throw new Error(`NIP44: unknown encryption version: ${data[0]}`)
+  }
+
+  const nonce = data.slice(1, 25)
+  const ciphertext = data.slice(25)
+  const plaintext = xchacha20(key, nonce, ciphertext)
+
+  return utf8Decoder.decode(plaintext)
+}
+```
+
+## Kotlin
+
+```kotlin
+// implementation 'fr.acinq.secp256k1:secp256k1-kmp-jni-android:0.10.1'
+// implementation "com.goterl:lazysodium-android:5.1.0@aar"
+// implementation "net.java.dev.jna:jna:5.12.1@aar"
+
+fun getSharedSecretNIP44(privKey: ByteArray, pubKey: ByteArray): ByteArray =
+  MessageDigest.getInstance("SHA-256").digest(
+    Secp256k1.get().pubKeyTweakMul(
+      Hex.decode("02") + pubKey,
+      privKey
+    ).copyOfRange(1, 33)
+  )
+
+fun encryptNIP44(msg: String, privKey: ByteArray, pubKey: ByteArray): String {
+  val nonce = ByteArray(24).apply {
+    SecureRandom.getInstanceStrong().nextBytes(this)
+  }
+
+  val cipher = streamXChaCha20Xor(
+    message = msg.toByteArray(),
+    nonce = nonce,
+    key = getSharedSecretNIP44(privKey, pubKey)
+  )
+
+  return Base64.getEncoder().encodeToString(
+    byteArrayOf(Nip44Version.XChaCha20.code) + nonce + cipher
+  )
+}
+
+fun decryptNIP44(base64: String, privKey: ByteArray, pubKey: ByteArray): String? {
+  val byteArray = Base64.getDecoder().decode(base64)
+
+  require(byteArray[0].toInt() == Nip44Version.XChaCha20.code) { "NIP44: unknown encryption version" }
+
+  return streamXChaCha20Xor(
+    message = byteArray.copyOfRange(25, byteArray.size),
+    nonce = byteArray.copyOfRange(1, 25),
+    key = getSharedSecretNIP44(privKey, pubKey)
+  )?.decodeToString()
+}
+
+// This method is not exposed in AndroidSodium yet, but it will be in the next version.
+fun streamXChaCha20Xor(message: ByteArray, nonce: ByteArray, key: ByteArray): ByteArray? {
+  return with (SodiumAndroid()) {
+    val resultCipher = ByteArray(message.size)
+
+    val isSuccessful = crypto_stream_chacha20_xor_ic(
+      resultCipher,
+      message,
+      message.size.toLong(),
+      nonce.drop(16).toByteArray(), // chacha nonce is just the last 8 bytes.
```

Comment:
How does this work?

Reply:
Not on libsodium. That code is from them.

Reply:
https://github.com/jedisct1/libsodium/blob/4dc02ce8414e70b8960bea77abf4fe46efb99e93/src/libsodium/crypto_stream/xchacha20/stream_xchacha20.c#L45

Reply:
yeah, it's ok, same as in js then.